File: sem iii/NM/Nm lab/FixedIteration_lab.py
import math

# Lab 4 - Fixed Point Iteration Method
# f(x) = 3x - cos(x) - 1

def f(x):
    return 3*x - math.cos(x) - 1

# x = g(x)
def g(x):
    return (math.cos(x) + 1) / 3


def fixed_point(xn, eps, n):
    i = 0

    while i < n:
        # compute next approximation
        x_next = g(xn)

        # compute error
        error = abs(x_next - xn)

        print(f"Iteration {i+1}: x = {x_next:.5f}, error = {error:.5f}")

        # stopping condition
        if error < eps:
            return x_next

        # update xn
        xn = x_next
        i += 1

    return xn


if __name__ == '__main__':
    print("Program to find the root of the given function using fixed-point iteration method.")
    x = float(input("Enter the initial guess: "))
    eps = float(input("Enter the error tolerance value: "))
    itr = int(input("Enter the number of iterations: "))

    root = fixed_point(x, eps, itr)
    print(f"\nThe root of the function is {root:.5f}")

File: sem iii/NM/Nm lab/ho/FixedIteration_lab.c
#include <stdio.h>
#include <math.h>

// f(x) = 3x - cos(x) - 1
double f(double x) {
    return 3*x - cos(x) - 1;
}

// x = g(x)
double g(double x) {
    return (cos(x) + 1) / 3;
}

int main() {
    double x, x_new, error, eps;
    int itr, i = 0;

    printf("Program to find root using Fixed Point Iteration Method\n");
    printf("Enter the initial guess: ");
    scanf("%lf", &x);
    printf("Enter the error tolerance: ");
    scanf("%lf", &eps);
    printf("Enter the number of iterations: ");
    scanf("%d", &itr);

    while (i < itr) {
        x_new = g(x);
        error = fabs(x_new - x);

        printf("Iteration %d: x = %.5lf  error = %.5lf\n",
               i+1, x_new, error);

        if (error < eps)
            break;

        x = x_new;
        i++;
    }

    printf("\nThe root of the function is %.5lf\n", x_new);
    return 0;
}
